File: backend/app/schemas/kb_settings.py
"""KB Settings Pydantic schemas for KB-level configuration.

Story 7.12: Typed Pydantic schemas for KB-level configuration with runtime
validation and type safety throughout the codebase.
"""

from enum import Enum
from typing import ClassVar
from uuid import UUID

from pydantic import BaseModel, Field

# =============================================================================
# Enum Types (Task 1: AC #1, #2, #7, #8)
# =============================================================================


class ChunkingStrategy(str, Enum):
    """Strategy for document chunking."""

    FIXED = "fixed"
    RECURSIVE = "recursive"
    SEMANTIC = "semantic"


class RetrievalMethod(str, Enum):
    """Method for document retrieval."""

    VECTOR = "vector"
    HYBRID = "hybrid"
    HYDE = "hyde"


class CitationStyle(str, Enum):
    """Style for citation formatting."""

    INLINE = "inline"
    FOOTNOTE = "footnote"
    NONE = "none"


class UncertaintyHandling(str, Enum):
    """How to handle uncertain/low-confidence responses."""

    ACKNOWLEDGE = "acknowledge"
    REFUSE = "refuse"
    BEST_EFFORT = "best_effort"


class TruncationStrategy(str, Enum):
    """Strategy for truncating text that exceeds max length."""

    START = "start"
    END = "end"
    NONE = "none"


class PoolingStrategy(str, Enum):
    """Strategy for pooling embeddings."""

    MEAN = "mean"
    CLS = "cls"
    MAX = "max"
    LAST = "last"


class DocumentParserBackend(str, Enum):
    """Document parser backend selection for all supported formats.

    Story 7.32: Docling Document Parser Integration
    - UNSTRUCTURED: Current default parser (Unstructured library)
    - DOCLING: Advanced parser with better table extraction and layout analysis
    - AUTO: Try Docling first, fallback to Unstructured on failure
    """

    UNSTRUCTURED = "unstructured"  # Current default
    DOCLING = "docling"  # Advanced parser (requires LUMIKB_PARSER_DOCLING_ENABLED=true)
    AUTO = "auto"  # Try docling, fallback to unstructured


# =============================================================================
# Sub-Config Schemas (Task 2: AC #1-8)
# =============================================================================


class ChunkingConfig(BaseModel):
    """Configuration for document chunking (AC-7.12.1)."""

    strategy: ChunkingStrategy = Field(default=ChunkingStrategy.RECURSIVE)
    chunk_size: int = Field(default=512, ge=100, le=2000)
    chunk_overlap: int = Field(default=50, ge=0, le=500)
    separators: list[str] = Field(default_factory=lambda: ["\n\n", "\n", " ", ""])

    model_config = {"extra": "forbid"}


class RetrievalConfig(BaseModel):
    """Configuration for document retrieval (AC-7.12.2)."""

    top_k: int = Field(default=10, ge=1, le=100)
    similarity_threshold: float = Field(default=0.7, ge=0.0, le=1.0)
    method: RetrievalMethod = Field(default=RetrievalMethod.VECTOR)
    mmr_enabled: bool = Field(default=False)
    mmr_lambda: float = Field(default=0.5, ge=0.0, le=1.0)
    hybrid_alpha: float = Field(default=0.5, ge=0.0, le=1.0)

    model_config = {"extra": "forbid"}


class RerankingConfig(BaseModel):
    """Configuration for result reranking (AC-7.12.3)."""

    enabled: bool = Field(default=False)
    model: str | None = Field(default=None)
    top_n: int = Field(default=10, ge=1, le=50)

    model_config = {"extra": "forbid"}


class GenerationConfig(BaseModel):
    """Configuration for text generation (AC-7.12.4)."""

    temperature: float = Field(default=0.7, ge=0.0, le=2.0)
    top_p: float = Field(default=0.9, ge=0.0, le=1.0)
    top_k: int = Field(default=40, ge=1, le=100)
    max_tokens: int = Field(default=2048, ge=100, le=16000)
    frequency_penalty: float = Field(default=0.0, ge=-2.0, le=2.0)
    presence_penalty: float = Field(default=0.0, ge=-2.0, le=2.0)
    stop_sequences: list[str] = Field(default_factory=list)

    model_config = {"extra": "forbid"}


class NERConfig(BaseModel):
    """Configuration for Named Entity Recognition (AC-7.12.5)."""

    enabled: bool = Field(default=False)
    confidence_threshold: float = Field(default=0.8, ge=0.0, le=1.0)
    entity_types: list[str] = Field(default_factory=list)
    batch_size: int = Field(default=32, ge=1, le=100)

    model_config = {"extra": "forbid"}


class DocumentProcessingConfig(BaseModel):
    """Configuration for document processing (AC-7.12.6, AC-7.32.2).

    Story 7.32: Added parser_backend field for Docling integration.
    """

    ocr_enabled: bool = Field(default=False)
    language_detection: bool = Field(default=True)
    table_extraction: bool = Field(default=True)
    image_extraction: bool = Field(default=False)
    parser_backend: DocumentParserBackend = Field(
        default=DocumentParserBackend.UNSTRUCTURED,
        description=(
            "Document parser backend for all supported formats (PDF, DOCX, MD). "
            "Requires LUMIKB_PARSER_DOCLING_ENABLED=true for 'docling' or 'auto' options."
        ),
    )

    model_config = {"extra": "forbid"}


class KBPromptConfig(BaseModel):
    """Configuration for KB-specific prompts (AC-7.12.7)."""

    system_prompt: str = Field(default="", max_length=4000)
    context_template: str = Field(default="")
    citation_style: CitationStyle = Field(default=CitationStyle.INLINE)
    uncertainty_handling: UncertaintyHandling = Field(
        default=UncertaintyHandling.ACKNOWLEDGE
    )
    response_language: str = Field(default="en")

    model_config = {"extra": "forbid"}


class EmbeddingConfig(BaseModel):
    """Configuration for embedding generation (AC-7.12.8, AC-7.12.11)."""

    model_id: UUID | None = Field(default=None)
    batch_size: int = Field(default=32, ge=1, le=100)
    normalize: bool = Field(default=True)
    truncation: TruncationStrategy = Field(default=TruncationStrategy.END)
    max_length: int = Field(default=512, ge=128, le=16384)
    prefix_document: str = Field(default="", max_length=100)
    prefix_query: str = Field(default="", max_length=100)
    pooling_strategy: PoolingStrategy = Field(default=PoolingStrategy.MEAN)

    model_config = {"extra": "forbid"}

    # Re-indexing fields that invalidate existing vectors (AC-7.12.11)
    REINDEX_FIELDS: ClassVar[set[str]] = {
        "model_id",
        "normalize",
        "prefix_document",
        "prefix_query",
        "pooling_strategy",
    }

    def requires_reindex(self, previous: "EmbeddingConfig | None") -> bool:
        """Check if changes require document re-indexing.

        Args:
            previous: Previous embedding configuration to compare against.

        Returns:
            True if any re-indexing field changed, False otherwise.
        """
        if previous is None:
            return False
        for field in self.REINDEX_FIELDS:
            if getattr(self, field) != getattr(previous, field):
                return True
        return False

    def detect_reindex_fields(self, previous: "EmbeddingConfig | None") -> list[str]:
        """Return list of changed fields that require re-indexing.

        Args:
            previous: Previous embedding configuration to compare against.

        Returns:
            List of field names that changed and require re-indexing.
        """
        if previous is None:
            return []
        return [
            f for f in self.REINDEX_FIELDS if getattr(self, f) != getattr(previous, f)
        ]

    def get_reindex_warning(self, changed_fields: list[str]) -> str:
        """Generate user-friendly warning about re-indexing requirement.

        Args:
            changed_fields: List of field names that changed.

        Returns:
            Warning message string, or empty string if no fields changed.
        """
        if not changed_fields:
            return ""
        field_names = ", ".join(changed_fields)
        return (
            f"Changing {field_names} requires re-indexing all documents "
            "in this knowledge base. This operation may take significant time "
            "depending on the number of documents."
        )


# =============================================================================
# Composite KBSettings Schema (Task 3: AC #9-10)
# =============================================================================


class KBSettings(BaseModel):
    """Complete KB-level configuration stored in settings JSONB (AC-7.12.9).

    This schema aggregates all sub-configuration sections with default
    factories. Empty {} settings parse with all defaults applied (AC-7.12.10).

    The three-layer precedence is:
    - Request params (highest priority)
    - KB settings (this schema)
    - System defaults (lowest priority)
    """

    chunking: ChunkingConfig = Field(default_factory=ChunkingConfig)
    retrieval: RetrievalConfig = Field(default_factory=RetrievalConfig)
    reranking: RerankingConfig = Field(default_factory=RerankingConfig)
    generation: GenerationConfig = Field(default_factory=GenerationConfig)
    ner: NERConfig = Field(default_factory=NERConfig)
    processing: DocumentProcessingConfig = Field(
        default_factory=DocumentProcessingConfig
    )
    prompts: KBPromptConfig = Field(default_factory=KBPromptConfig)
    embedding: EmbeddingConfig = Field(default_factory=EmbeddingConfig)
    preset: str | None = Field(
        default=None,
        description="Preset name used for configuration (e.g., 'legal', 'technical')",
    )
    debug_mode: bool = Field(
        default=False,
        description=(
            "Enable debug mode for RAG pipeline telemetry. When enabled, "
            "chat responses include detailed information about retrieved chunks, "
            "similarity scores, KB parameters, and timing metrics."
        ),
    )

    model_config = {"extra": "forbid"}


# =============================================================================
# Preset Schemas (Story 7.16: KB Settings Presets)
# =============================================================================


class PresetInfo(BaseModel):
    """Information about a single KB settings preset (AC-7.16.1).

    Used for listing available presets in the UI dropdown.
    """

    id: str = Field(description="Preset identifier (e.g., 'legal', 'technical')")
    name: str = Field(description="Display name for the preset")
    description: str = Field(description="Brief description of the preset's purpose")


class PresetListResponse(BaseModel):
    """Response schema for listing all available presets (AC-7.16.1)."""

    presets: list[PresetInfo] = Field(
        description="List of available KB settings presets"
    )


class PresetDetailResponse(BaseModel):
    """Response schema for getting a single preset with full settings (AC-7.16.2-6)."""

    id: str = Field(description="Preset identifier (e.g., 'legal', 'technical')")
    name: str = Field(description="Display name for the preset")
    description: str = Field(description="Brief description of the preset's purpose")
    settings: "KBSettings" = Field(
        description="Full settings configuration for this preset"
    )


class PresetDetectRequest(BaseModel):
    """Request schema for detecting which preset matches given settings (AC-7.16.8)."""

    chunking: ChunkingConfig = Field(default_factory=ChunkingConfig)
    retrieval: RetrievalConfig = Field(default_factory=RetrievalConfig)
    generation: GenerationConfig = Field(default_factory=GenerationConfig)
    prompts: KBPromptConfig = Field(default_factory=KBPromptConfig)
    preset: str | None = Field(default=None)


class PresetDetectResponse(BaseModel):
    """Response schema for preset detection (AC-7.16.8)."""

    detected_preset: str = Field(
        description="Detected preset ID ('legal', 'technical', etc.) or 'custom' if no match"
    )

File: frontend/src/types/kb-settings.ts
/**
 * KB Settings TypeScript types - Story 7-14: KB Settings UI - General Panel
 *
 * Mirrors backend schemas from backend/app/schemas/kb_settings.py
 * for type-safe KB configuration management.
 */

// =============================================================================
// Enum Types (mirroring backend enums)
// =============================================================================

export enum ChunkingStrategy {
  FIXED = 'fixed',
  RECURSIVE = 'recursive',
  SEMANTIC = 'semantic',
}

export enum RetrievalMethod {
  VECTOR = 'vector',
  HYBRID = 'hybrid',
  HYDE = 'hyde',
}

export enum CitationStyle {
  INLINE = 'inline',
  FOOTNOTE = 'footnote',
  NONE = 'none',
}

export enum UncertaintyHandling {
  ACKNOWLEDGE = 'acknowledge',
  REFUSE = 'refuse',
  BEST_EFFORT = 'best_effort',
}

export enum TruncationStrategy {
  START = 'start',
  END = 'end',
  NONE = 'none',
}

export enum PoolingStrategy {
  MEAN = 'mean',
  CLS = 'cls',
  MAX = 'max',
  LAST = 'last',
}

/**
 * Document parser backend selection for all supported formats.
 * Story 7-32: Docling Document Parser Integration
 */
export enum DocumentParserBackend {
  UNSTRUCTURED = 'unstructured',
  DOCLING = 'docling',
  AUTO = 'auto',
}

// =============================================================================
// Sub-Config Interfaces (mirroring backend Pydantic models)
// =============================================================================

/**
 * Configuration for document chunking (AC-7.14.2)
 * Mirrors ChunkingConfig in backend
 */
export interface ChunkingConfig {
  strategy: ChunkingStrategy;
  chunk_size: number; // 100-2000, default 512
  chunk_overlap: number; // 0-500, default 50
  separators?: string[];
}

/**
 * Configuration for document retrieval (AC-7.14.3)
 * Mirrors RetrievalConfig in backend
 */
export interface RetrievalConfig {
  top_k: number; // 1-100, default 10
  similarity_threshold: number; // 0.0-1.0, default 0.7
  method: RetrievalMethod;
  mmr_enabled: boolean;
  mmr_lambda: number; // 0.0-1.0, default 0.5
  hybrid_alpha?: number; // 0.0-1.0, default 0.5
}

/**
 * Configuration for result reranking
 * Mirrors RerankingConfig in backend
 */
export interface RerankingConfig {
  enabled: boolean;
  model: string | null;
  top_n: number; // 1-50, default 10
}

/**
 * Configuration for text generation (AC-7.14.4)
 * Mirrors GenerationConfig in backend
 */
export interface GenerationConfig {
  temperature: number; // 0.0-2.0, default 0.7
  top_p: number; // 0.0-1.0, default 0.9
  top_k?: number; // 1-100, default 40
  max_tokens: number; // 100-16000, default 2048
  frequency_penalty?: number; // -2.0-2.0, default 0.0
  presence_penalty?: number; // -2.0-2.0, default 0.0
  stop_sequences?: string[];
}

/**
 * Configuration for Named Entity Recognition
 * Mirrors NERConfig in backend
 */
export interface NERConfig {
  enabled: boolean;
  confidence_threshold: number; // 0.0-1.0, default 0.8
  entity_types: string[];
  batch_size: number; // 1-100, default 32
}

/**
 * Configuration for document processing
 * Mirrors DocumentProcessingConfig in backend
 * Story 7-32: Added parser_backend field
 */
export interface DocumentProcessingConfig {
  ocr_enabled: boolean;
  language_detection: boolean;
  table_extraction: boolean;
  image_extraction: boolean;
  parser_backend: DocumentParserBackend;
}

/**
 * Configuration for KB-specific prompts
 * Mirrors KBPromptConfig in backend
 */
export interface KBPromptConfig {
  system_prompt: string; // max 4000 chars
  context_template: string;
  citation_style: CitationStyle;
  uncertainty_handling: UncertaintyHandling;
  response_language: string;
}

/**
 * Configuration for embedding generation
 * Mirrors EmbeddingConfig in backend
 */
export interface EmbeddingConfig {
  model_id: string | null;
  batch_size: number; // 1-100, default 32
  normalize: boolean;
  truncation: TruncationStrategy;
  max_length: number; // 128-16384, default 512
  prefix_document: string; // max 100 chars
  prefix_query: string; // max 100 chars
  pooling_strategy: PoolingStrategy;
}

// =============================================================================
// Composite KBSettings Type (AC-7.14.6, AC-7.14.8)
// =============================================================================

/**
 * Complete KB-level configuration stored in settings JSONB
 * Mirrors KBSettings in backend
 */
export interface KBSettings {
  chunking: ChunkingConfig;
  retrieval: RetrievalConfig;
  reranking: RerankingConfig;
  generation: GenerationConfig;
  ner: NERConfig;
  processing: DocumentProcessingConfig;
  prompts: KBPromptConfig;
  embedding: EmbeddingConfig;
  preset: string | null;
  /**
   * Enable debug mode for RAG pipeline telemetry (AC-9.15.1).
   * When enabled, chat responses include detailed information about
   * retrieved chunks, similarity scores, KB parameters, and timing metrics.
   */
  debug_mode: boolean;
}

// =============================================================================
// Default Values (matching backend defaults)
// =============================================================================

export const DEFAULT_CHUNKING_CONFIG: ChunkingConfig = {
  strategy: ChunkingStrategy.RECURSIVE,
  chunk_size: 512,
  chunk_overlap: 50,
  separators: ['\n\n', '\n', ' ', ''],
};

export const DEFAULT_RETRIEVAL_CONFIG: RetrievalConfig = {
  top_k: 10,
  similarity_threshold: 0.7,
  method: RetrievalMethod.VECTOR,
  mmr_enabled: false,
  mmr_lambda: 0.5,
  hybrid_alpha: 0.5,
};

export const DEFAULT_RERANKING_CONFIG: RerankingConfig = {
  enabled: false,
  model: null,
  top_n: 10,
};

export const DEFAULT_GENERATION_CONFIG: GenerationConfig = {
  temperature: 0.7,
  top_p: 0.9,
  top_k: 40,
  max_tokens: 2048,
  frequency_penalty: 0.0,
  presence_penalty: 0.0,
  stop_sequences: [],
};

export const DEFAULT_NER_CONFIG: NERConfig = {
  enabled: false,
  confidence_threshold: 0.8,
  entity_types: [],
  batch_size: 32,
};

export const DEFAULT_PROCESSING_CONFIG: DocumentProcessingConfig = {
  ocr_enabled: false,
  language_detection: true,
  table_extraction: true,
  image_extraction: false,
  parser_backend: DocumentParserBackend.UNSTRUCTURED,
};

export const DEFAULT_PROMPT_CONFIG: KBPromptConfig = {
  system_prompt: '',
  context_template: '',
  citation_style: CitationStyle.INLINE,
  uncertainty_handling: UncertaintyHandling.ACKNOWLEDGE,
  response_language: 'en',
};

export const DEFAULT_EMBEDDING_CONFIG: EmbeddingConfig = {
  model_id: null,
  batch_size: 32,
  normalize: true,
  truncation: TruncationStrategy.END,
  max_length: 512,
  prefix_document: '',
  prefix_query: '',
  pooling_strategy: PoolingStrategy.MEAN,
};

export const DEFAULT_KB_SETTINGS: KBSettings = {
  chunking: DEFAULT_CHUNKING_CONFIG,
  retrieval: DEFAULT_RETRIEVAL_CONFIG,
  reranking: DEFAULT_RERANKING_CONFIG,
  generation: DEFAULT_GENERATION_CONFIG,
  ner: DEFAULT_NER_CONFIG,
  processing: DEFAULT_PROCESSING_CONFIG,
  prompts: DEFAULT_PROMPT_CONFIG,
  embedding: DEFAULT_EMBEDDING_CONFIG,
  preset: null,
  debug_mode: false,
};

// =============================================================================
// API Types for Settings Endpoints (AC-7.14.8)
// =============================================================================

/**
 * Response from GET /api/v1/knowledge-bases/{id}/settings
 */
export type KBSettingsResponse = KBSettings;

/**
 * Request body for PUT /api/v1/knowledge-bases/{id}/settings
 * All fields are optional for partial updates
 */
export interface KBSettingsUpdate {
  chunking?: Partial<ChunkingConfig>;
  retrieval?: Partial<RetrievalConfig>;
  reranking?: Partial<RerankingConfig>;
  generation?: Partial<GenerationConfig>;
  ner?: Partial<NERConfig>;
  processing?: Partial<DocumentProcessingConfig>;
  prompts?: Partial<KBPromptConfig>;
  embedding?: Partial<EmbeddingConfig>;
  preset?: string | null;
  debug_mode?: boolean;
}

// =============================================================================
// Validation Constraints (for form validation)
// =============================================================================

export const KB_SETTINGS_CONSTRAINTS = {
  chunking: {
    chunk_size: { min: 100, max: 2000 },
    chunk_overlap: { min: 0, max: 500 },
  },
  retrieval: {
    top_k: { min: 1, max: 100 },
    similarity_threshold: { min: 0.0, max: 1.0 },
    mmr_lambda: { min: 0.0, max: 1.0 },
    hybrid_alpha: { min: 0.0, max: 1.0 },
  },
  generation: {
    temperature: { min: 0.0, max: 2.0 },
    top_p: { min: 0.0, max: 1.0 },
    max_tokens: { min: 100, max: 16000 },
    frequency_penalty: { min: -2.0, max: 2.0 },
    presence_penalty: { min: -2.0, max: 2.0 },
  },
} as const;

// =============================================================================
// Display Labels for UI
// =============================================================================

export const CHUNKING_STRATEGY_LABELS: Record<ChunkingStrategy, string> = {
  [ChunkingStrategy.FIXED]: 'Fixed',
  [ChunkingStrategy.RECURSIVE]: 'Recursive',
  [ChunkingStrategy.SEMANTIC]: 'Semantic',
};

export const RETRIEVAL_METHOD_LABELS: Record<RetrievalMethod, string> = {
  [RetrievalMethod.VECTOR]: 'Vector',
  [RetrievalMethod.HYBRID]: 'Hybrid',
  [RetrievalMethod.HYDE]: 'HyDE',
};

export const CITATION_STYLE_LABELS: Record<CitationStyle, string> = {
  [CitationStyle.INLINE]: 'Inline',
  [CitationStyle.FOOTNOTE]: 'Footnote',
  [CitationStyle.NONE]: 'None',
};

export const UNCERTAINTY_HANDLING_LABELS: Record<UncertaintyHandling, string> = {
  [UncertaintyHandling.ACKNOWLEDGE]: 'Acknowledge Uncertainty',
  [UncertaintyHandling.REFUSE]: 'Refuse Response',
  [UncertaintyHandling.BEST_EFFORT]: 'Best Effort',
};

/**
 * Display labels for DocumentParserBackend enum
 * Story 7-32: Docling Document Parser Integration (AC-7.32.3)
 */
export const DOCUMENT_PARSER_BACKEND_LABELS: Record<DocumentParserBackend, string> = {
  [DocumentParserBackend.UNSTRUCTURED]: 'Unstructured (Default)',
  [DocumentParserBackend.DOCLING]: 'Docling',
  [DocumentParserBackend.AUTO]: 'Auto',
};

export const DOCUMENT_PARSER_BACKEND_DESCRIPTIONS: Record<DocumentParserBackend, string> = {
  [DocumentParserBackend.UNSTRUCTURED]: 'Standard parser for PDF, DOCX, Markdown',
  [DocumentParserBackend.DOCLING]: 'Advanced parser with better tables and layout analysis',
  [DocumentParserBackend.AUTO]: 'Try Docling first, fallback to Unstructured on failure',
};
